src/storage/src/table/streaming_table/state_table.rs

```diff
@@ -612,12 +625,21 @@ impl<S: StateStore> StateTable<S> {
         &mut self,
         buffer: BTreeMap<Vec<u8>, RowOp>,
         epoch: u64,
+        watermark: Option<&ScalarImpl>,
     ) -> StorageResult<()> {
         let mut write_batch = self.keyspace.start_write_batch(WriteOptions {
             epoch,
             table_id: self.table_id(),
         });
+
+        let prefix_serializer = self.pk_serde.prefix(1);
+        let range_end_suffix = watermark.map(|watermark| {
+            serialize_pk(&Row::new(vec![Some(watermark.clone())]), &prefix_serializer)
+        });
         for (pk, row_op) in buffer {
+            if let Some(ref range_end) = range_end_suffix && &pk[1..] < range_end.as_slice() {
```

Comment:
Why `pk[1..]`? I think its just `pk[..]`?

Reply:
> Why `pk[1..]`? I think its just `pk[..]`?

`pk[0]` is vnode

Reply:
Indeed, can you use `VIRTUAL_NODE_SIZE` instead of `1`?

Reply:
> Indeed, can you use `VIRTUAL_NODE_SIZE` instead of `1`?

good catch. thank you.